Documentation/design-docs/code-versioning-proposal.md

```diff
@@ -0,0 +1,146 @@
+# Code Versioning #
+
+The current runtime prototype for tiered compilation doesn't work properly when trying to use profiler
+rejit at the same time. Although this may seem initially seem like a corner case intersection, diagnostic monitoring profilers are increasingly becoming standard operating procedure with rejit usage increasing fast in that domain. Likewise the goal is for tiered compilation to be used as an always on (or nearly always on) feature. Rejit can also be viewed as a canonical example of a broader problem that the runtime doesn't have a foundational concept of code versioning and lifetime management. Rejit has one conception of it, the tiered jitting prototype adds another, Edit and Continue adds a 3rd, LCG a 4th, the IL interpreter a 5th, Samsung's in progress code-pitching a 6th, and so on. Using any of those features is mutually exclusive with every other at method level granularity.
+
+I don't aim to produce in one effort a grand unified implementation, I do want to achieve two things:
+
+1. A roadmap that allows all those features to interoperate if we made the effort to do so in the future.
+2. Implement that design far enough so that Rejit + Tiered Compilation interoperate now.
+
+For now tiered compilation can reasonably co-exist with the other features by disabling it only on the methods that are eligible for one of the other techniques.
```

Comment:
So, that means a profiler which does IL-Rewriting that uses Rejit, cannot work with Tiered Compilation? How would it not work? Would RequestReJIT fail? Or would it fail already when COR_PRF_ENABLE_REJIT is requested?

If COR_PRF_ENABLE_REJIT is not requested, Tiered Compilation works just fine?

How would Tiered Compilation be disabled? Is there a way for the profiler to do it (via bitmask)? Can it be disabled just for certain methods and how? What mode is used if it's disabled? Would it always produce highest-quality code (resulting in most CPU usage)?